pkg/transparentproxy/iptables/builder/builder_restore.go

```diff
@@ -87,79 +87,115 @@ func (e executable) exec(ctx context.Context, args ...string) (*bytes.Buffer, er
 
-	if e.save.path == "" {
-		missing = append(missing, e.save.name)
+	if e.Save.Path == "" {
+		missing = append(missing, e.Save.Name)
 	}
 
-	if e.restore.path == "" {
-		missing = append(missing, e.restore.name)
+	if e.Restore.Path == "" {
+		missing = append(missing, e.Restore.Name)
 	}
 
 	if len(missing) > 0 {
-		return errors.Errorf("couldn't find executables: [%s]", strings.Join(missing, ","))
+		return nil, errors.Errorf("couldn't find executables: [%s]", strings.Join(missing, ","))
 	}
 
 	// We always need to have access to the "nat" table
-	if stdout, err := e.save.exec(ctx, "-t", "nat"); err != nil {
-		return errors.Wrap(err, "couldn't verify if table: 'nat' is available")
+	if stdout, err := e.Save.exec(ctx, "-t", "nat"); err != nil {
+		return nil, errors.Wrap(err, "couldn't verify if table: 'nat' is available")
 	} else if cfg.ShouldRedirectDNS() || cfg.ShouldCaptureAllDNS() {
 		e.foundDockerOutputChain = dockerOutputChainRegex.Match(stdout.Bytes())
 	}
 
-	if cfg.ShouldConntrackZoneSplit() {
-		if _, err := e.save.exec(ctx, "-t", "raw"); err != nil {
-			return errors.Wrap(err, "couldn't verify if table: 'raw' is available")
+	// It seems in some cases (GKE with ContainerOS), even if "iptables-nft" is available
+	// there are some kernel modules with iptables match extensions missing.
+	for _, matchExtension := range necessaryMatchExtensions {
+		if _, err := e.Iptables.exec(ctx, "-m", matchExtension, "--help"); err != nil {
+			return nil, errors.Wrapf(err, "verification if match: %q exist failed", matchExtension)
+		}
+	}
+
+	if cfg.ShouldConntrackZoneSplit(e.Iptables.Path) {
+		if _, err := e.Save.exec(ctx, "-t", "raw"); err != nil {
+			return nil, errors.Wrap(err, "couldn't verify if table: 'raw' is available")
 		}
 	}
 
-	return nil
+	return e, nil
 }
 
-func detectIptablesExecutables(ctx context.Context, cfg config.Config, ipv6 bool) (*executables, bool, error) {
-	nft := newExecutables(ipv6, "nft")
-	legacy := newExecutables(ipv6, "legacy")
+func (e *Executables) withFallback(fallback *Executables) *Executables {
+	if fallback != nil {
+		e.fallback = fallback
+	}
+
+	return e
+}
+
+func DetectIptablesExecutables(
+	ctx context.Context,
+	cfg config.Config,
+	ipv6 bool,
+) (*Executables, error) {
+	nft, nftVerifyErr := newExecutables(ipv6, "nft").verify(ctx, cfg)
+	legacy, legacyVerifyErr := newExecutables(ipv6, "legacy").verify(ctx, cfg)
+
+	if nftVerifyErr != nil && legacyVerifyErr != nil {
+		return nil, fmt.Errorf("no valid iptables executable found: %s, %s", nftVerifyErr, legacyVerifyErr)
+	}
 
-	if err := nft.verify(ctx, cfg); err != nil {
-		return legacy, true, legacy.verify(ctx, cfg)
+	if nftVerifyErr != nil {
+		return legacy, nil
 	}
 
 	// Found DOCKER_OUTPUT chain in iptables-nft
 	if nft.foundDockerOutputChain {
-		return nft, false, nil
+		return nft.withFallback(legacy), nil
 	}
 
-	if err := legacy.verify(ctx, cfg); err != nil {
-		return nft, false, nil
+	if legacyVerifyErr != nil {
```

Comment:
shouldn't this be above the `if nft.foundDockerOutputChain` statement? like we can set fallback to legacy and have `legacyVerifyErr` right?

Reply:
not relevant (see https://github.com/kumahq/kuma/pull/9776/files/74f853bfea5d1cde6743176f7acd15b0741ab42f#r1547940971)